docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,278 @@
+```
+
+pull multiple files and show their manifests as pretty JSON in standard output.
+
+```shell
+$ oras pull $REGISTRY/$REPO:$TAG --format '{{jsonPretty .}}'
+```
+
+```json
+{
+    "reference": "$REGISTRY/$REPO:$DIGEST",
+    "files" : [
+        {
+            "mediaType": "application/vnd.oci.image.layer.v1.tar",
+            "digest": "sha256:d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+            "size": 12,
+            "annotations": {
+                "org.opencontainers.image.created":"2023-11-29T06:32:43Z"
+            },
+            "path":"path1/artifact1.json"
+            "reference": "$REGISTRY/$REPO:$layer0_digest"
+        },
+        {
+            "mediaType": "application/vnd.oci.image.layer.v1.tar",
+            "digest": "sha256:4add5a911ba64df27458ed8229da804a26d2a84f4b8b650937ec8f73cd8be2c7",
+            "size": 12,
+            "annotations": {
+                "org.opencontainers.image.created":"2023-11-30T06:32:43Z"
+            },
+            "path":"path2/artifact2.json"
+            "reference": "$REGISTRY/$REPO:$layer1_digest"
+        }
+    ]
+}
+```
+
+### oras attach
+
+Attach an artifact to an image and show the manifest of the attached file in JSON format.
+
+```shell
+$ oras attach --artifact-type example/sbom $REGISTRY/$REPO:$TAG sbom.spdx --format json {{ toPrettyJson }}
+```
+
+```json
+{
+    "mediaType": "application/vnd.oci.image.manifest.v1+json",
+    "digest": "sha256:d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+    "size": 12,
+    "annotations": {
+        "org.opencontainers.image.created":"2023-11-29T06:32:43Z"
+    },
+    "artifactType" : "example/sbom",
+    "reference": "$REGISTRY/$REPO:$digest"
+}
+```
+
+### oras push
+
+Push an artifact to a repository and show the metadata of the pushed artifact.
+
+```shell
+$ oras push $REGISTRY/$REPO:$TAG --format '{{jsonPretty .}}'
+```
+
+```json
+{
+    "mediaType": "application/vnd.oci.image.layer.v1.tar",
+    "digest": "sha256:d2a84f4b8b650937ec8f73cd8be2c74add5a911ba64df27458ed8229da804a26",
+    "size": 12,
+    "annotations": {
+        "org.opencontainers.image.title": "hello.txt"
+    },
+    "artifactType": "application/vnd.example+type",
+    "reference": "$REGISTRY/$REPO:$digest"
+}
+```
+
+### oras discover
+
+Discover an artifact's referrers. The default output should be listed in a tree view.
+
+```shell
+$ oras discover localhost:5000/hello:v1
+```
+
+```
+localhost:5000/hello/demo@sha256:04beb34cd24389147b4642a828b47fabefa722dea794dc3834567cf014ab0fe6
+└── application/vnd.oci.empty.v1+json
+    ├── sha256:1b82e249d83eb4881b8bf4ff9cf13a28799907ddc624b4c3c9140fa77d54fa42
+    ├── sha256:28653e2bb5b5a75393c3a8b58ed9998796299b41dc1ff1f55b9f0844ad7ba39c
+```
+
+Discover an artifact's referrers manifest in JSON. Consider `oras discover` is more likely used by terminal (human) than machine, the default formatted output with `--format json` should be pretty JSON.
```

Comment:
It is possible to support `--format json|table|tree|$TEMPLATE`. In that case, `--format` can covers all usages of  `--output` and we should consider deprecating `--output`.

I suggest in the release we introduce `--format` (planned v1.2.0)
1) add a deprecate mark to `--output` flag in the help doc
2) If user use `--output json|table|tree`, we print a warning in stderr to suggest them to use `--format json|table|tree` instead
3) deprecate `--output` in v2.0

Reply:
> Also --output should be in v2.0
Do you mean deprecate ` --output` should be in ORAS v2.0?

Reply:
Added a note about the deprecation of the `--output` flag in v1.2.0